sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala

```diff
@@ -84,9 +84,16 @@ object ResolveDefaultColumns extends QueryErrorsBase
     if (SQLConf.get.enableDefaultColumns) {
       val newFields: Seq[StructField] = tableSchema.fields.map { field =>
         if (field.metadata.contains(CURRENT_DEFAULT_COLUMN_METADATA_KEY)) {
-          val analyzed: Expression = analyze(field, statementType)
+          val defaultSql: String = if (field.dataType.isInstanceOf[VariantType]) {
+            // A variant's SQL/string representation is its JSON string which cannot be directly
+            // casted to a variant type. Thus, we lazily evaluate the default expression to avoid
```

Comment:
Is there a literal presentation for the variant type? The inconsistency here looks weird. We should have something like `variant{'key': 'value'}`

Reply:
hmm, I'm not sure if we have a literal presentation for the variant type (its a struct of two binary fields), which is why we have this problem. Similar to the year month time interval/calendar interval, for example, we always rely on `parse_json` (which can accept a json string) to create a literal variant.

The variant's `toString` method essentially calls `toJson` which converts the encoded variant into a JSON string

Reply:
From the unit tests, can we store text like `parse_json('1')` in the default value?

Reply:
Yep, `parse_json('1')` works (even before this PR). This is because the current code implicitly inserts a cast to variant, which is ok in this case because `CAST(1 as VARIANT)` succeeds

In the more complex case such as a default of `parse_json('{'k': 'v'}')`, the analyzer actually fails because it tries to analyze the sql text `{'k': 'v'}` (note the lack of outer quotations). An alternative here is to create a string literal from  `{'k': 'v'}` and wrap it with `parse_json` (effectively coercing it to the correct variant type). This feels inefficient, however, because we have variant->string->variant conversion rather than lazily evaluating it once

Reply:
> analyze the sql text {'k': 'v'}

shouldn't it be sql text `parse_json('{'k': 'v'}')`? I am confusing...

> create a string literal from {'k': 'v'} and wrap it with parse_json (effectively coercing it to the correct variant type). This feels inefficient, however, because we have variant->string->variant conversion rather than lazily evaluating it once

Since the `analyze` will perform `ConstantFolding`, it would be great if we can unify the code path here. The overhead is trivial here.

Reply:
> shouldn't it be sql text parse_json('{'k': 'v'}')? I am confusing...

ah sorry, I skipped a few steps: when we create a table (or alter table add column) and add a column default, `constantFoldCurrentDefaultsToExistDefaults` [here](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala#L81) is called to constant fold the expression and save it as a sql expression string under the `EXISTS_DEFAULT_COLUMN_METADATA_KEY` metadata key.

So for `parse_json('{'k': 'v'}')` the string `{'k': 'v'}` (no quotes) is saved under the `EXISTS_DEFAULT_COLUMN_METADATA_KEY` key. Later on, in `getExistenceDefaultValues` [here](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala#L387) we retrieve the sql string in `EXISTS_DEFAULT_COLUMN_METADATA_KEY` and run `analyze` on it again to get the actual sql expression used as the default value. Now, `EXISTS_DEFAULT_COLUMN_METADATA_KEY` is `{'k': 'v'}` (no outer quotations) and throws an analysis error.

> it would be great if we can unify the code path here

Happy to be wrong, but I personally think this approach is actually cleaner than the alternative (wrapping `{'k': 'v'}` with parse_json). This is because the alternative would require us to do something like:
```
val defaultSql: String = if (field.dataType.isInstanceOf[VariantType]) {
            analyzed match {
              case l: ExprLiteral if l.value == null => l.sql
              // A variant's SQL/string representation is its JSON string which cannot be directly
              // casted to a variant type. Thus, `analyzed` is wrapped with `parse_json` so the
              // default value is transformed into the correct variant.
              case l: ExprLiteral =>
                ParseJson(ExprLiteral(UTF8String.fromString(analyzed.sql), StringType)).sql
              case c: Cast => c.sql
            }
          } else {
            analyzed.sql
       }
```

```
val defaultSql: String = if (field.dataType.isInstanceOf[VariantType]) {
            // A variant's SQL/string representation is its JSON string which cannot be directly
            // casted to a variant type. Thus, we lazily evaluate the default expression to avoid
            // materializing a string representation of the variant.
            field.metadata.getString(CURRENT_DEFAULT_COLUMN_METADATA_KEY)
          } else {
            analyze(field, statementType).sql
          }
```

i.e. we would rely on `VariantValue.toString` to always return a json string (which may or may not change in the future), wrap that as a string literal, and input it to `parse_json`.